### ISSUE TITLE
No difference between adversarial and original samples after evasion attack

### ISSUE THREAD

**Issue by willi3by** [Trusted-AI/adversarial-robustness-toolbox#902]
**Describe the bug**
I am testing adversarial attacks on a trained 3D ResNet-50 binary classifier. So far, I have attempted the Carlini l2 and PGDNumpy attacks, as well as tried to calculate the CLEVER metric. However, I keep getting what I think are unexpected results for each method. For the attacks, the loss actually decreases and accuracy remains the same as the test set as I increase eps. Since this is the size of the perturbation, I would expect accuracy to get worse (hopefully generating a realistic security curve). For the CLEVER score, not matter what I set the radius to, the score is always 0. When investigating the issue, I found that the generated adversarial examples are exactly the same as the original test data, which I imagine is leading to all of these issues. I have 125 samples of data that are 91x109x91x1 and scaled to values between 0 and 1. I am using the R interface to Tensorflow and Keras, and reticulate for all python interaction. Here is the code for the network:

```
build_ResNet <- function(input_shape, num_classes, loss = "binary_crossentropy", optimizer = optimizer_adam(lr=0.001), metrics){
  if(num_classes < 3){
    num_units = 1
    activation = "sigmoid"
  }
  else {
    num_units = num_classes
    activation = "softmax"
    }

  model <- keras_model_sequential() %>%
    layer_conv_3d(filters = 64, kernel_size = c(7,7,7), strides = c(2,2,2), padding = 'same',
                  use_bias = F, input_shape = input_shape) %>%
    layer_batch_normalization() %>%
    layer_activation_relu() %>%
    layer_max_pooling_3d(pool_size = c(3,3,3), strides = c(2,2,2), padding = 'same')

  prev_filters <<- 64
  filter_list <<- c(rep(64,3), rep(128, 4), rep(256,6), rep(512,3))
  for(i in seq_along(filter_list)){
    i <<- i
    if(filter_list[i] == prev_filters){strides <<- 1} else {strides <<- 2}
    model %>% residual_layer(filters = filter_list[i], strides = strides)
    prev_filters <<- filter_list[i]
  }

  model <- model %>%
    layer_global_average_pooling_3d() %>%
    layer_flatten() %>%
    layer_dropout(0.3) %>%
    layer_dense(units = num_units, activation = activation)

  model <- model %>% compile(loss="binary_crossentropy",
                    optimizer = optimizer,
                    metrics = metrics)

  return(model)
}
```

```
tf$compat$v1$disable_eager_execution()
model <- build_ResNet(input_shape = c(91,109,91,1), 
                      num_classes = 2, optimizer = optimizer_adam(lr=0.0001), 
                      metrics = list("accuracy"))

model <- model %>% load_model_weights_hdf5('./model/best_model.h5')
```

**To Reproduce**
Here are snippets that I am using to test the two attack methods:
```
classifier <- art$estimators$classification$KerasClassifier(model=model, clip_values = c(0,1))
cw_attack <- art$attacks$evasion$CarliniL2Method(classifier = classifier, learning_rate = 0.01, max_iter = 100, batch_size = 32)
x_test_adv <- cw_attack$generate(x_test, t(as.matrix(y_test))) #Only testing first 10 examples for now.
adv_diff <- mean(abs(x_test_adv - x_test)) #This is equal to 0.
```
```
pgd_attack <- art$attacks$evasion$ProjectedGradientDescentNumpy(estimator = classifier, 
                                                                  norm = 2, eps = 0.1, eps_step = 0.1,
                                                                  max_iter = 100L)
x_test_adv <- pgd_attack$generate(x = x_test, y = t(as.matrix(y_test)))
adv_diff <- mean(abs(x_test_adv - x_test)) #This is equal to 0.
```

**Expected behavior**
I would expect there to be a difference between the adversarial samples and the original samples. 

**System information (please complete the following information):**
 - OS: Windows 10, build 18363.1316
 - Python version: 3.7.9
 - ART version or commit number: 1.5.1
 - TensorFlow: 2.3.1 (Using Keras through tensorflow.keras)

**Comment by beat-buesser** [comment]
Hi @willi3by Thank you very much for using ART and providing a detailed description! My first question, does the binary classifier have a single output (`num_units = 1`) in these observations?

**Comment by willi3by** [comment]
Hi,

Thanks for the quick response! I am very much enjoying the toolbox and it is assisting me greatly in my first attempts to do robustness analyses. 

Yes, that is correct. The last layer is a dense layer with 1 unit and sigmoid activation. 

Thanks

**Comment by beat-buesser** [comment]
I'm very happy to hear that ART is useful for you!

I think the single output classifier is the reason. ART supports for classification either one-hot encoded labels or index labels, but assumes that the number of outputs by the model corresponds to the number of classes. For the single output, binary classifier current ART versions will see the model as always predicting class 0 and treat single column labels as class label 0, therefore the model is always predicting correct, which is misleading.

With the current versions of ART, there is a workaround by making the binary classifier predict 2 outputs (one for each class, in the case of predicting probabilities the two would sum up to 1) and provide the labels of shape (n_samples, 2).

We have already a few open issues on single output binary classifiers and we have recognised that there is great interest among the ART users in support for single output binary classifiers and we are planning to provide direct support with ART 1.7.

**Comment by willi3by** [comment]
Thanks for the info! I was able to load my model, set weights except for the last layer, and change the dense layer to have 2 units, and everything works as expected. Looking forward to the next release. Thanks again for your time.
